### PR TITLE
Use ES6 Modules

### PR DESCRIPTION
**Pull request by apmorton**
### TODO
- [x] mock-fs doesn't appear to work correctly, causing one of the tests to fail
- [x] decide on and normalize module exports
  - some modules `export default ClassName`
  - some modules `export { ClassName }`
  - some do both
- [x] check for required nodejs in `etc/scripts/find-node`
  - we now absolutely require v12.18
    - I don't remember the specifics, but there was something added during the 12.x series that is required
- [x] if possible change code to avoid any remaining dynamic imports
  - look at `lib/tooling/index.js` as an example alternate design
    - TL;DR; statically import all possible implementations and use object lookup to get class instead of dynamically importing
    - downside is that we statically import all implementations, including potentially unused ones - although so long as modules don't have side effects this should be a non-issue
    - upside is that we avoid extra async init stuff that is otherwise problematic in some places (like in constructors)
- [ ] refactor modules to avoid side effects and other top level "stuff"
  - modules should ideally only declare (and export) classes, constants, and naked functions.
  - es6 imports all occur before any of your own module top level code is executed, so this makes it impossible to have the kind of control some of our old designs relied on

### Module Export Format
According to [this article](https://medium.com/@timoxley/named-exports-as-the-default-export-api-670b1b554f65) the current state-of-the-art for ES6 is to simply export your names and _not_ have a default export at all.
I tend to agree based on what I noticed during the refactoring process.

consider:
```javascript
// thing.js
export default class Thing { };

// main.js
import Stuff from './thing`;
```

versus:
```javascript
// thing.js
export class Thing { };

// main.js
import { Thing as Stuff } from './thing`;
```

In the `export default` example it is not entirely obvious that `Stuff` is actually `Thing` - and nothing actually enforces that you use consistent naming when writing the imports.

In the `export` example, you can still get the same result if desired, but you have to be explicit about it.  If you just `import { Thing } from './thing';` and then change the class name in `thing.js` the import breaks and you must either propagate the name change to the imported code or consciously decide to alias it using `import { Thing as Stuff } from './thing';`

Additionally, I think I prefer this:
```javascript
export class Thing { };

export class Stuff { };
```

over:
```javascript
class Thing { };

class Stuff { };

export { Thing, Stuff };
```

and will try to refactor things around that instead.

### CODE REVIEW

**Comment by mattgodbolt** [comment]
Thanks @apmorton this is great! Especially the rationale for imports etc.

I cloned locally and got:
```
error: Exception while processing compilers:No demangler type named 'undefined' found {"stack":"Error: No demangler type named 'undefined' found\n    at getDemanglerTypeFromName (/home/matthew/dev/ce/compiler-explorer/lib/demangler/index.js:23:15)\n    at new BaseCompiler (/home/matthew/dev/ce/compiler-explorer/lib/base-compiler.js:78:35)\n    at new Win32Compiler (/home/matthew/dev/ce/compiler-explorer/lib/compilers/win32.js:32:8)\n    at new Win32VcCompiler (/home/matthew/dev/ce/compiler-explorer/lib/compilers/win32-vc.js:31:9)\n    at CompileHandler.create (/home/matthew/dev/ce/compiler-explorer/lib/handlers/compile.js:134:20)\n    at /home/matthew/dev/ce/compiler-explorer/node_modules/underscore/underscore.js:73:21\n    at Proxy.map (/home/matthew/dev/ce/compiler-explorer/node_modules/underscore/underscore.js:207:24)\n    at CompileHandler.setCompilers (/home/matthew/dev/ce/compiler-explorer/lib/handlers/compile.js:142:68)\n    at CompilerFinder.find (/home/matthew/dev/ce/compiler-explorer/lib/compiler-finder.js:396:47)\n    at processTicksAndRejections (internal/process/task_queues.js:94:5)"}
info: Configured with Sentry endpoint https://8e4614f649ad4e3faf3e7e8827b935f9@sentry.io/102028
error: #### No compilers found: no compilation will be done!
```

but I suspect that's a specific local properties I have. Will dig a bit more so I can play about for a bit.

But I love the way this is goin: thanks.

**Comment by mattgodbolt** [comment]
(confirmed, moving my local c++.local.properties away "fixed" it, but that's a copy of the current prod c++ properties. maybe once conflicts are resolved we'll be ok)

**Comment by apmorton** [comment]
yeah - `demanglerClassFile` changed to `demanglerType` (to match the pattern we have with `compilerType`), which is a breaking change

**Comment by partouf** [comment]
> yeah - `demanglerClassFile` changed to `demanglerType` (to match the pattern we have with `compilerType`), which is a breaking change

Next time with these refactoring jobs I would advise not to implement all the changes in the world at once. The ES6 thing is already breaking enough.

We should also probably have a way to somehow communicate breaking changes to users that have local installations with custom compilers and things. Maybe a dedicated thing in docs?

**Comment by apmorton** [comment]
This PR is definitely going to be interesting to merge because of the sheer scale - but I don't know that there was much to be done about that.

The breaking change is due to eliminating dynamic import statements, which are async in ES6 where require statements would not have been.

In some places it was easy enough to throw a `await import(x)`, but it was not possible in others without other refactoring. I decided to just eliminate all the dynamic imports (besides stuff for using webpack in development) for consistency sake.

I may well split up this PR where possible to make it less painful to merge.

Perhaps the demangler config change is a good first candidate - we can land that now and document the change in preparation for ES6

**Comment by apmorton** [comment]
@mattgodbolt @partouf @RabsRincon I think this is ready for merge - but its a massive set of changes, so please give it a thorough look.

Also, if possible, can we try to freeze master until this is merged?  I would like to avoid any more rebases on this branch 😩  (unless someone finds an issue during review that needs more time to be fixed)

I would like to have at least two reviews on this before merging.

**Review by RabsRincon — APPROVED** [review]

**Comment by mattgodbolt** [comment]
Just made master "push via PRs only" for now as a "freeze"....looking now!

File: lib/compilers/argument-parsers.js [hunk 1]
```diff
@@ -91,14 +92,11 @@ class GCCParser extends BaseParser {
...
             // -masm= may be available but unsupported by the compiler.
-            // eslint-disable-next-line promise/catch-or-return
-            compiler.exec(compiler.compiler.exe, ['-fsyntax-only', '--target-help', '-masm=intel'])
-                .then(result => {
-                    if (result.code === 0) {
-                        compiler.compiler.intelAsm = '-masm=intel';
-                        compiler.compiler.supportsIntel = true;
-                    }
-                });
+            const res = await compiler.exec(compiler.compiler.exe, ['-fsyntax-only', '--target-help', '-masm=intel']);
```
Comment (mattgodbolt) [inline lib/compilers/argument-parsers.js:95]: nice, thanks for fixing this up while you were here

File: lib/keyed-type.js [hunk 2]
```diff
@@ -36,7 +36,7 @@ function makeKeyMap(typeName, objects) {
     return keyToTypeMap;
 }
 
-function makeKeyedTypeGetter(typeName, objects) {
+export function makeKeyedTypeGetter(typeName, objects) {
```
Comment (mattgodbolt) [inline lib/keyed-type.js:39]: neat solution btw

File: lib/shortener/base.js [hunk 3]
```diff
@@ -1,4 +1,4 @@
-// Copyright (c) 2018, Compiler Explorer Authors
+// Copyright (c) 2020, Compiler Explorer Authors
```
Comment (mattgodbolt) [inline lib/shortener/base.js:1]: no biggie, I thought we were leaving copyrights alone? I forget how we left it :)

**Comment by partouf** [comment]
Tested under Windows to just see if everything still worked. Looks good aside from the breakage due to the previous compilerType merge.

Linter works fine under Windows as well.

One thing I noticed with `npm run test`, is that I now know the reason for the path.resolves, because those tests are now broken under Windows:

```
  4) CompilerDropInTool
       Should support gcc based compilers:

      AssertionError: expected [ Array(2) ] to deeply equal [ Array(2) ]
      + expected - actual

       [
      -  "--gcc-toolchain=D:\\opt\\compiler-explorer\\gcc-8.0"
      -  "--gcc-toolchain=D:\\opt\\compiler-explorer\\gcc-8.0"
      +  "--gcc-toolchain=/opt/compiler-explorer/gcc-8.0"
      +  "--gcc-toolchain=/opt/compiler-explorer/gcc-8.0"
       ]

      at Context.<anonymous> (test\tool-tests.js:66:33)
      at processImmediate (internal/timers.js:456:21)
      at process.topLevelDomainCallback (domain.js:137:15)

  5) CompilerDropInTool
       Should support ICC compilers:

      AssertionError: expected [ Array(2) ] to deeply equal [ Array(2) ]
      + expected - actual

       [
      -  "--gcc-toolchain=D:\\opt\\compiler-explorer\\gcc-8.2.0"
      -  "--gcc-toolchain=D:\\opt\\compiler-explorer\\gcc-8.2.0"
      +  "--gcc-toolchain=/opt/compiler-explorer/gcc-8.2.0"
      +  "--gcc-toolchain=/opt/compiler-explorer/gcc-8.2.0"
       ]
```

More tests are broken under Windows, so we can worry about this a later time.

**Review by mattgodbolt — APPROVED** [review]
* Can we add a line to AddingALanguage.md and the other docs to say about adding it to compilers/`_all.js` please? I know it was missing before.. (and if there's other docs I've missed, a quick one there too)

**Comment by mattgodbolt** [comment]
Thanks @partouf - yeah I think we can delay fixing those issues if @apmorton is happy to merge as-is (though I'll file a bug once it's merged to fix them, and probably try and fix myself)

**Pull request merged**